File: 222.count-complete-tree-nodes.py
#
# @lc app=leetcode id=222 lang=python3
#
# [222] Count Complete Tree Nodes
#

# Time:  O((logn)^2)
# Space: O(1)


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def countNodes(self, root: TreeNode) -> int:
        left, lh, right, rh = root, 0, root, 0
        while left:
            left, lh = left.left, lh + 1
        while right:
            right, rh = right.right, rh + 1
        if lh == rh:
            return 2 ** lh - 1
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)


# @lc code=end

File: 222.count-complete-tree-nodes.js
/*
 * @lc app=leetcode id=222 lang=javascript
 *
 * [222] Count Complete Tree Nodes
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @return {number}
 */
var countNodes = function (root) {
  let left = root;
  let right = root;
  let leftH = 0;
  let rightH = 0;
  while (left) {
    left = left.left;
    ++leftH;
  }
  while (right) {
    right = right.right;
    ++rightH;
  }
  if (leftH === rightH) {
    return 2 ** leftH - 1;
  }
  return 1 + countNodes(root.left) + countNodes(root.right);
};
// @lc code=end
